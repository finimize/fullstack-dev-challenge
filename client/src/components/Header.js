import React from 'react'

export const Header = () => (
  <div className="header-banner">
    <h1 className="fmz-white-font">Finimize Interest Rate Calculator</h1>
  </div>
)
