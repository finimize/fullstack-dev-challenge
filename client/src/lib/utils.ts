import type { ChartTooltipItem } from 'chart.js'
import { SAVINGS_DEFAULTS } from './constants'

export const createQueryString = (queries: { [key: string]: any }): string =>
    Object.keys(queries)
        .map((key) => encodeURIComponent(key) + '=' + encodeURIComponent(queries[key]))
        .join('&')

export const appendYearsQuery = (queryString: string) =>
    `${queryString}&years=${SAVINGS_DEFAULTS.YEARS}`

export const extractYearlyData = (projections: number[]) =>
    projections.filter((item, index) => index % 12 === 0)

export const numberFormatter = new Intl.NumberFormat('gb-GB', {
    style: 'currency',
    currency: 'GBP',
})

export const getTooltipTitle = (tooltipItems: ChartTooltipItem[]) => `Year ${tooltipItems[0].label}`

export const getTooltipLabel = (tooltipItem: ChartTooltipItem) => {
    const value = tooltipItem.value ? parseFloat(tooltipItem.value) : 0
    return numberFormatter.format(value)
}
