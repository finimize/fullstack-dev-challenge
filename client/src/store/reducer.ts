import {
    UPDATE_DETAILS,
    UPDATE_CALCULATIONS,
    UPDATE_INTEREST,
    UPDATE_DATA,
    VALIDATE_DETAILS,
    NEXT_PAGE,
    TOGGLE_MODE,
    PREVIOUS_PAGE,
} from './types'
import {
    StateInterface,
    UpdateDetailsTypeInterface,
    UpdateCalculationsTypeInterface,
    UpdateInterestRateInterface,
    UpdateDataInterface,
    ValidateDetailsTypeInterface,
    NextPageTypeInterface,
    ToggleModeInterface,
    PreviousPageTypeInterface,
} from './store.interface'

export const reducer = (
    initialState: StateInterface,
    action:
        | UpdateDetailsTypeInterface
        | UpdateCalculationsTypeInterface
        | UpdateInterestRateInterface
        | UpdateDataInterface
        | ValidateDetailsTypeInterface
        | NextPageTypeInterface
        | PreviousPageTypeInterface
        | ToggleModeInterface,
): StateInterface => {
    switch (action.type) {
        case NEXT_PAGE: {
            return {
                ...initialState,
                currentPage: initialState.currentPage + 1,
            }
        }
        case PREVIOUS_PAGE: {
            return {
                ...initialState,
                currentPage: initialState.currentPage - 1,
            }
        }
        case TOGGLE_MODE: {
            return {
                ...initialState,
                calculatorMode: action.payload,
            }
        }
        case UPDATE_DETAILS:
            return {
                ...initialState,
                details: {
                    ...initialState.details,
                    [action.field]: action.payload,
                },
            }
        case UPDATE_CALCULATIONS:
            return {
                ...initialState,
                calculations: {
                    ...initialState.calculations,
                    [action.field]: action.payload,
                },
            }
        case UPDATE_INTEREST:
            return {
                ...initialState,
                calculations: {
                    ...initialState.calculations,
                    interestRate: action.payload,
                },
            }
        case UPDATE_DATA:
            return {
                ...initialState,
                data: {
                    ...initialState.data,
                    value: action.payload,
                },
            }
        case VALIDATE_DETAILS:
            return {
                ...initialState,
                details: {
                    ...initialState.details,
                    firstName: {
                        ...initialState.details.firstName,
                        error: action.payload.firstName,
                    },
                    lastName: {
                        ...initialState.details.lastName,
                        error: action.payload.lastName,
                    },
                    email: {
                        ...initialState.details.email,
                        error: action.payload.email,
                    },
                },
            }
        default:
            return initialState
    }
}
