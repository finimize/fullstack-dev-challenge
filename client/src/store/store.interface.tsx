import {
    UPDATE_DETAILS,
    UPDATE_CALCULATIONS,
    UPDATE_INTEREST,
    UPDATE_DATA,
    VALIDATE_DETAILS,
    NEXT_PAGE,
    TOGGLE_MODE,
    TOGGLE_COMPOUNDING_FREQUENCY,
    PREVIOUS_PAGE,
} from './types'

export interface TextInputInterface {
    value: string
    error: string
}

interface DetailsInterface {
    firstName: TextInputInterface
    lastName: TextInputInterface
    email: TextInputInterface
}

export interface DataValueInterface {
    yearlySavings: number[]
    finalValue: number
    yearlyBreakdown: {
        yearFinal: number
        savings: []
    }[]
}

export interface DataInterface {
    value: DataValueInterface | null
    error: string
    isLoading: boolean
}

export interface CalculationsInterface {
    initialSavings: TextInputInterface
    interestRate: number
    compoundingFrequency: number
    monthlyContributions: TextInputInterface
}

export interface StateInterface {
    details: DetailsInterface
    data: DataInterface
    currentPage: number
    calculatorMode: string
    calculations: CalculationsInterface
}

export type UpdateDetailsTypeInterface = {
    type: typeof UPDATE_DETAILS
    field: 'firstName' | 'lastName' | 'email'
    payload: Partial<TextInputInterface>
}
export type UpdateCalculationsTypeInterface = {
    type: typeof UPDATE_CALCULATIONS
    field: 'initialSavings' | 'monthlyContributions'
    payload: Partial<TextInputInterface>
}

export type UpdateInterestRateInterface = {
    type: typeof UPDATE_INTEREST
    payload: number
}

export type AppProps = {
    children: React.ReactNode
}

export type ValidateDetailsTypeInterface = {
    type: typeof VALIDATE_DETAILS
    payload: {
        firstName: string
        lastName: string
        email: string
    }
}

export type NextPageTypeInterface = {
    type: typeof NEXT_PAGE
}

export type PreviousPageTypeInterface = {
    type: typeof PREVIOUS_PAGE
}

export type ToggleModeInterface = {
    type: typeof TOGGLE_MODE
    payload: string
}

export enum CompoundingFrequency {
    annually = 1,
    quarterly = 4,
    monthly = 12,
}

export type ToggleCompoundingFrequencyInterface = {
    type: typeof TOGGLE_COMPOUNDING_FREQUENCY
    payload: string
}

export type UpdateDataInterface = {
    type: typeof UPDATE_DATA
    payload: DataInterface
}
