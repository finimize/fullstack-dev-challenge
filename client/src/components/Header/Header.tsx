import React, { FC, useContext } from 'react'
import { Box, Text } from '@chakra-ui/react'
import { StepButton } from '../StepButton'
import { AppContext } from '../../store'
import { GOT_TO_PAGE } from '../../store/types'
import { useValidateDetails } from '../Hooks/useValidateDetails'

export const Header: FC = () => {
    const { state, dispatch } = useContext(AppContext)
    const validationData = useValidateDetails()

    return (
        <Box backgroundColor='white' padding='6'>
            <Box display='flex' width='100%' justifyContent='space-evenly' padding='6'>
                {Array.from({ length: 3 }).map((e, i) => (
                    <StepButton
                        key={i}
                        label={i}
                        disabled={i > 0 && !validationData.valid}
                        selected={state.currentPage === i + 1}
                        onClick={() =>
                            dispatch({
                                type: GOT_TO_PAGE,
                                payload: i + 1,
                            })
                        }
                    />
                ))}
            </Box>
            <Text
                fontSize='xxl'
                marginLeft='8'
                marginBottom='4'
                color='blueHeader'
                fontWeight='500'
            >
                Let&apos;s start the savings...
            </Text>
        </Box>
    )
}
