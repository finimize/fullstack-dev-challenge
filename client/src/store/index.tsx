import React, { createContext, useReducer, FC } from 'react'
import { initialState } from './initialState'
import { reducer } from './reducer'
import {
    UpdateDetailsTypeInterface,
    ValidateDetailsTypeInterface,
    NextPageTypeInterface,
    StateInterface,
    AppProps,
    PreviousPageTypeInterface,
    ToggleModeInterface,
    UpdateCalculationsTypeInterface,
    UpdateInterestRateInterface,
} from './store.interface'

const AppContext = createContext<{
    state: StateInterface
    dispatch: React.Dispatch<
        | UpdateDetailsTypeInterface
        | UpdateCalculationsTypeInterface
        | ValidateDetailsTypeInterface
        | NextPageTypeInterface
        | ToggleModeInterface
        | PreviousPageTypeInterface
        | UpdateInterestRateInterface
    >
}>({
    state: initialState,
    dispatch: () => null,
})

const AppProvider: FC<AppProps> = ({ children }) => {
    const [state, dispatch] = useReducer(reducer, initialState)
    return <AppContext.Provider value={{ state, dispatch }}>{children}</AppContext.Provider>
}

export { AppContext, AppProvider }
