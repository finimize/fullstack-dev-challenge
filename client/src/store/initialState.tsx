import { StateInterface, CompoundingFrequency } from './store.interface'

export const initialState: StateInterface = {
    details: {
        firstName: {
            value: '',
            error: '',
        },
        lastName: {
            value: '',
            error: '',
        },
        email: {
            value: '',
            error: '',
        },
    },
    data: {
        value: null,
        error: '',
        isLoading: false,
    },
    calculations: {
        initialSavings: {
            value: '10',
            error: '',
        },
        interestRate: 20,
        compoundingFrequency: CompoundingFrequency.annually,
        monthlyContributions: {
            value: '0',
            error: '',
        },
    },
    calculatorMode: 'Simple',
    currentPage: 1,
}
