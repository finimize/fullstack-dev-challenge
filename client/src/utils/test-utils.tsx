import React, { FC, useReducer } from 'react'
import { ChakraProvider, extendTheme } from '@chakra-ui/react'
import { render as rtlRender, RenderResult, RenderOptions } from '@testing-library/react'
import { AppContext, DispatchType } from '../store'
import { initialState } from '../store/initialState'
import { reducer } from '../store/reducer'
import { StateInterface } from '../store/store.interface'
import { theme } from '../theme'

const mainTheme = extendTheme(theme)

interface CustomRenderOptions extends Partial<RenderOptions> {
    store?: StateInterface
    defaultTheme?: typeof theme
}

export interface CustomRenderInterface {
    (
        ui: React.ReactElement<unknown, string | React.JSXElementConstructor<unknown>>,
        options: CustomRenderOptions,
    ): { render: RenderResult; dispatchMock: jest.Mock<unknown, unknown[]> }
}

const customRender: CustomRenderInterface = (
    ui,
    { store = initialState, defaultTheme = mainTheme, ...options },
) => {
    const dispatchMock = jest.fn()
    const AllProviders: FC = ({ children }) => {
        // const [calculatorState, setCalculatorState] = useState(store)
        const [calculatorState, dispatch] = useReducer(reducer, store)
        return (
            <ChakraProvider theme={defaultTheme}>
                <AppContext.Provider
                    value={{
                        state: calculatorState,
                        dispatch: dispatchMock.mockImplementation((value: DispatchType) =>
                            dispatch(value),
                        ),
                    }}
                >
                    {' '}
                    {children}{' '}
                </AppContext.Provider>{' '}
            </ChakraProvider>
        )
    }
    return { render: rtlRender(ui, { wrapper: AllProviders, ...options }), dispatchMock }
}

export { customRender }
