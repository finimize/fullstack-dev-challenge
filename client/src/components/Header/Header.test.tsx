import React from 'react'
import { cleanup } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { initialState } from '../../store/initialState'
import { customRender } from '../../utils/test-utils'
import { Header } from './Header'

afterEach(cleanup)

describe('component/Savings Controls', () => {
    it('renders Header component', () => {
        const { render } = customRender(<Header />, {})
        const { container } = render
        expect(container).toMatchSnapshot()
    })
    it('renders Header component', () => {
        const { render } = customRender(<Header />, {})
        const { getByTestId } = render
        const step1 = getByTestId('step-button-0')
        const step2 = getByTestId('step-button-1')
        const step3 = getByTestId('step-button-2')
        expect(step1).not.toBeDisabled()
        expect(step2).toBeDisabled()
        expect(step3).toBeDisabled()
    })

    it('renders Header component', () => {
        const { render, dispatchMock } = customRender(<Header />, {
            store: {
                ...initialState,
                details: {
                    firstName: {
                        value: 'john smith',
                        error: '',
                    },
                    lastName: {
                        value: 'john smith',
                        error: '',
                    },
                    email: {
                        value: 'johnsmith@gmail.com',
                        error: '',
                    },
                },
            },
        })
        const { getByTestId } = render
        const step1 = getByTestId('step-button-0')
        const step2 = getByTestId('step-button-1')
        const step3 = getByTestId('step-button-2')
        expect(step1).not.toBeDisabled()
        expect(step2).not.toBeDisabled()
        expect(step3).not.toBeDisabled()
        userEvent.click(step2)
        expect(dispatchMock).toBeCalledWith({ payload: 2, type: 'GOT_TO_PAGE' })
        userEvent.click(step3)
        expect(dispatchMock).toBeCalledWith({ payload: 3, type: 'GOT_TO_PAGE' })
    })
})
