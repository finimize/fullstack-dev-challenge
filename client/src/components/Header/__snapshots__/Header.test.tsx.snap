// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/Savings Controls renders Header component 1`] = `
<div>
   
  <div
    class="css-yf46qw"
  >
    <div
      class="css-1rb18jj"
    >
      <button
        class="chakra-button css-10dvod4"
        data-testid="step-button-0"
        type="button"
      >
        0
      </button>
      <button
        class="chakra-button css-1qos78r"
        data-testid="step-button-1"
        disabled=""
        type="button"
      >
        1
      </button>
      <button
        class="chakra-button css-1qos78r"
        data-testid="step-button-2"
        disabled=""
        type="button"
      >
        2
      </button>
    </div>
    <p
      class="chakra-text css-1wvvwig"
    >
      Let's start the savings...
    </p>
  </div>
   
   
</div>
`;
